ee/src/main/java/org/jboss/as/ee/structure/InitializeInOrderProcessor.java

```diff
@@ -49,24 +50,40 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
         if (parent == null) {
             return;
         }
+
         final EarMetaData earConfig = parent.getAttachment(Attachments.EAR_METADATA);
-        if (earConfig != null) {
-            final boolean inOrder = earConfig.getInitializeInOrder();
-            if (inOrder && earConfig.getModules().size() > 1) {
+        final List<String> startupBeanModules = deploymentUnit.getParent().getAttachmentList(org.jboss.as.ee.component.Attachments.DEPLOYMENT_STARTUP_MODULES);
+        final Map<String, DeploymentUnit> deploymentUnitMap = new HashMap<String, DeploymentUnit>();
+        final ResourceRoot deploymentRoot = deploymentUnit.getAttachment(org.jboss.as.server.deployment.Attachments.DEPLOYMENT_ROOT);
+        final ModuleMetaData thisModulesMetadata = deploymentRoot.getAttachment(Attachments.MODULE_META_DATA);
 
-                final Map<String, DeploymentUnit> deploymentUnitMap = new HashMap<String, DeploymentUnit>();
-                for (final DeploymentUnit subDeployment : parent.getAttachment(org.jboss.as.server.deployment.Attachments.SUB_DEPLOYMENTS)) {
+        if(earConfig != null && (earConfig.getInitializeInOrder() || (startupBeanModules!=null && !startupBeanModules.isEmpty())) && (startupBeanModules==null || !startupBeanModules.contains(deploymentUnit.getName()))) {
```

Comment:
`startupBeanModules` will not be null, so no need for null check.  It may be empty.